pkg/plugin/processor/builtin/unwrap_debezium.go

```diff
@@ -0,0 +1,347 @@
+		return nil, nil
+	default:
+		return nil, cerrors.Errorf("expected a map or a string, got %T", val)
+	}
+}
+
+func (u *unwrapDebezium) validateRecord(data opencdc.StructuredData) error {
+	var multiErr error
+	if _, ok := data[debeziumFieldAfter]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", debeziumFieldAfter))
+	}
+	if _, ok := data[debeziumFieldSource]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", debeziumFieldSource))
+	}
+	if _, ok := data[debeziumFieldOp]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", debeziumFieldOp))
+	}
+	// ts_ms and transaction can be empty
+	return multiErr
+}
+
+func (u *unwrapDebezium) unwrapMetadata(rec opencdc.Record, dbzRec opencdc.StructuredData) (opencdc.Metadata, error) {
+	var source map[string]string
+	for field, val := range dbzRec {
+		switch field {
+		case debeziumFieldAfter, debeziumFieldBefore, debeziumFieldOp:
+			continue // ignore
+		case debeziumFieldTimestamp:
+			tsMs, ok := val.(float64)
+			if !ok {
+				return nil, cerrors.Errorf("%s is not a float", debeziumFieldTimestamp)
+			}
+			readAt := time.UnixMilli(int64(tsMs))
+			rec.Metadata.SetReadAt(readAt)
+		case debeziumFieldSource:
+			// don't add prefix for source fields to be consistent with the
+			// behavior of the debezium converter in the SDK - it puts all
+			// metadata fields into the `source` field
+			source = u.flatten("", val)
+		default:
+			flattened := u.flatten("debezium."+field, val)
+			for k, v := range flattened {
+				rec.Metadata[k] = v
+			}
+		}
+	}
+
+	// source is added at the end to overwrite any other fields
+	for k, v := range source {
+		rec.Metadata[k] = v
+	}
+
+	return rec.Metadata, nil
+}
+
+func (u *unwrapDebezium) flatten(key string, val any) map[string]string {
+	var prefix string
+	if len(key) > 0 {
+		prefix = key + "."
+	}
+	switch val := val.(type) {
+	case map[string]any:
+		out := make(map[string]string)
+		for k1, v1 := range val {
+			for k2, v2 := range u.flatten(prefix+k1, v1) {
+				out[k2] = v2
+			}
+		}
+		return out
+	case nil:
+		return nil
+	case string:
+		return map[string]string{key: val}
+	default:
+		return map[string]string{key: fmt.Sprint(val)}
+	}
+}
+
+// convertOperation converts debezium operation to openCDC operation
+func (u *unwrapDebezium) convertOperation(op string) (opencdc.Operation, error) {
+	switch op {
+	case debeziumOpCreate:
+		return opencdc.OperationCreate, nil
+	case debeziumOpUpdate:
+		return opencdc.OperationUpdate, nil
+	case debeziumOpDelete:
+		return opencdc.OperationDelete, nil
+	case debeziumOpRead:
+		return opencdc.OperationSnapshot, nil
+	case debeziumOpUnset:
+		return opencdc.OperationUpdate, nil
+	}
+	return 0, cerrors.Errorf("%q is an invalid operation", op)
+}
+
+func (u *unwrapDebezium) unwrapKey(key opencdc.Data) opencdc.Data {
```

Comment:
I wonder if these `unwrap` processors can reuse some of the same utilities, iirc this is the same as in the kafka connect processor. Maybe a refactoring down the road, once we have all processors merged in the same branch.

Reply:
Yup, yup, definitely, we can do a bit of refactoring down the road. I've been intentionally creating individual PRs to keep them small.